multiversx_sdk/core/controllers/smart_contract_controller.py

```diff
@@ -0,0 +1,226 @@
+        ...
+
+
+class IQueryResponse(Protocol):
+    return_data: List[str]
+    return_code: str
+    return_message: str
+    gas_used: int
+
+    def get_return_data_parts(self) -> List[bytes]:
+        ...
+
+
+class INetworkConfig(Protocol):
+    chain_id: str
+
+
+class INetworkProvider(Protocol):
+    def get_network_config(self) -> INetworkConfig:
+        ...
+
+    def query_contract(self, query: IQuery) -> IQueryResponse:
+        ...
+
+
+class IAccount(Protocol):
+    address: IAddress
+
+    def sign(self, data: bytes) -> bytes:
+        ...
+
+
+class IAbi(Protocol):
+    def encode_endpoint_input_parameters(self, endpoint_name: str, values: List[Any]) -> List[bytes]:
+        ...
+
+    def encode_constructor_input_parameters(self, values: List[Any]) -> List[bytes]:
+        ...
+
+    def encode_upgrade_constructor_input_parameters(self, values: List[Any]) -> List[bytes]:
+        ...
+
+    def decode_endpoint_output_parameters(self, endpoint_name: str, encoded_values: List[bytes]) -> List[Any]:
+        ...
+
+
+class SmartContractController:
+    def __init__(self, network_provider: INetworkProvider, abi: Optional[IAbi] = None) -> None:
+        self.chain_id: Union[str, None] = None
+        self.provider = network_provider
+        self.abi = abi
+        self.factory: Union[SmartContractTransactionsFactory, None] = None
+        self.parser = SmartContractTransactionsOutcomeParser()
+        self.query_controller = SmartContractQueriesController(
+            query_runner=QueryRunnerAdapter(self.provider),
+            abi=self.abi
+        )
+        self.tx_computer = TransactionComputer()
+
+    def create_transaction_for_deploy(self,
+                                      sender: IAccount,
+                                      nonce: int,
+                                      bytecode: Union[Path, bytes],
+                                      gas_limit: int,
+                                      arguments: Sequence[Any] = [],
+                                      native_transfer_amount: int = 0,
+                                      is_upgradeable: bool = True,
+                                      is_readable: bool = True,
+                                      is_payable: bool = False,
+                                      is_payable_by_sc: bool = True) -> Transaction:
+        self._ensure_factory_is_initialized()
+
+        transaction = self.factory.create_transaction_for_deploy(  # type: ignore
+            sender=sender.address,
+            bytecode=bytecode,
+            gas_limit=gas_limit,
+            arguments=arguments,
+            native_transfer_amount=native_transfer_amount,
+            is_upgradeable=is_upgradeable,
+            is_readable=is_readable,
+            is_payable=is_payable,
+            is_payable_by_sc=is_payable_by_sc
+        )
+
+        transaction.nonce = nonce
+        transaction.signature = sender.sign(self.tx_computer.compute_bytes_for_signing(transaction))
```

Comment:
Maybe we can have a base class for controllers, which holds common functionality - e.g a private function to sign transactions (just an idea)?

Reply:
Will keep it in mind as it could be a good idea, but as of now I don't think it is necessary.